kyo-reactive-streams/shared/src/main/scala/kyo/interop/reactive-streams/StreamSubscription.scala

```diff
@@ -0,0 +1,255 @@
+            case DownstreamState.Requesting(requested, Absent) =>
+                // Someone requested, we offer right away
+                val accepted      = Math.min(requested, n)
+                val nextRequested = requested - accepted
+                sideEffect = () => IO(Result.success(accepted))
+                DownstreamState.Requesting(nextRequested, Absent)
+            case DownstreamState.Finished =>
+                // Downstream cancelled
+                sideEffect = () => IO(Result.fail(()))
+                DownstreamState.Finished
+            case other =>
+                sideEffect = () => IO(Result.success(0L))
+                other
+        }
+        sideEffect()
+    end offer
+
+    override def request(n: Long): Unit =
+        if n <= 0 then subscriber.onError(new IllegalArgumentException("non-positive subscription request"))
+        var sideEffect: () => Unit = () => ()
+        state.update {
+            case DownstreamState.Requesting(0L, Present(offered -> promise)) =>
+                val accepted      = Math.min(offered, n)
+                val nextRequested = n - accepted
+                val nextOfferred  = offered - accepted
+                sideEffect = () => promise.completeDiscard(Result.success(accepted))
+                DownstreamState.Requesting(nextRequested, Absent)
+            case DownstreamState.Requesting(requested, Absent) =>
+                val nextRequested = Math.min(Long.MaxValue - requested, n) + requested
+                sideEffect = () => ()
+                DownstreamState.Requesting(nextRequested, Absent)
+            case other =>
+                sideEffect = () => ()
+                other
+        }
+        sideEffect()
+    end request
+
+    override def cancel(): Unit =
+        given Frame                = Frame.internal
+        var sideEffect: () => Unit = () => ()
+        state.update {
+            case DownstreamState.Requesting(_, Present(_ -> promise)) =>
+                sideEffect = () => promise.completeDiscard(Result.fail(()))
+                DownstreamState.Finished
+            case other =>
+                sideEffect = () => ()
+                DownstreamState.Finished
+        }
+        sideEffect()
+    end cancel
+
+    def subscribe: Unit < (IO & Ctx) =
+        var sideEffect: () => Unit < (IO & Ctx) = () => IO.unit
+        state.update {
+            case DownstreamState.Uninitialized =>
+                sideEffect = () =>
+                    IO {
+                        subscriber.onSubscribe(this)
+                    }
+                DownstreamState.Requesting(0L, Absent)
+            case other =>
+                sideEffect = () => IO.unit
+                other
+        }
+        sideEffect()
+    end subscribe
+
+    private[reactivestreams] def consume(
+        using
+        tag: Tag[Emit[Chunk[V]]],
+        frame: Frame,
+        safepoint: Safepoint
+    ): Fiber[Nothing, StreamFinishState] < (IO & Ctx) =
```

Comment:
Basically, we need to fork this into a new fiber, which will run the stream until completion. Maybe Poll can help us eliminate the need to fork, which results in eliminating boundary

Reply:
Yeah, I can't see why forking is necessary. Wouldn't `Poll.run(stream.emit)(poll).map(_._2)` directly work?

Reply:
Other lib like ZIO and Fs2 also has to fork. AFAIK, it's mainly because draining stream is a "runForever" task until interrupted or stream is done. Without forking it will block other thing from being done in the current thread.
I'll try to figure out if I can eliminate it